crates/floresta-chain/src/util.rs

```diff
@@ -0,0 +1,641 @@
+use std::cmp::min;
+use std::ops::Add;
+
+use bitcoin::block::Header;
+use bitcoin::consensus::encode::serialize_hex;
+use bitcoin::BlockHash;
+use bitcoin::Work;
+
+use crate::BlockchainInterface;
+
+/// Provides additional methods for working with `Header` objects,
+/// extending their functionality to interact with blockchain data.
+pub trait HeaderExt {
+    /// Calculates the Median Time Past (MTP) for the block.
+    ///
+    /// The MTP is the median of the timestamps of the last 11 blocks,
+    /// or fewer if the block height is less than 11.
+    fn calculate_median_time_past(
+        &self,
+        chain: &impl BlockchainInterface,
+    ) -> Result<u32, HeaderExtError>;
+
+    /// Calculates the total chain work up to the current block.
+    ///
+    /// This sums the work for all blocks in the chain
+    fn calculate_chain_work(
+        &self,
+        chain: &impl BlockchainInterface,
+    ) -> Result<Work, HeaderExtError>;
+
+    /// Retrieves the hash of the next block in the chain, if it exists.
+    ///
+    /// Returns `None` if the block is the tip of the chain.
+    fn get_next_block_hash(
+        &self,
+        chain: &impl BlockchainInterface,
+    ) -> Result<Option<BlockHash>, HeaderExtError>;
+
+    /// Returns the block's "bits" field as a hexadecimal string.
+    fn get_bits(&self) -> String;
+
+    /// Calculates the total chain work up to the current block and
+    /// returns it as a hexadecimal string.
+    fn get_chain_work(&self, chain: &impl BlockchainInterface) -> Result<String, HeaderExtError>;
+
+    /// Calculates the number of confirmations for the block.
+    ///
+    /// The number of confirmations is the difference between the chain's
+    /// current height and the block's height, plus one.
+    fn get_confirmations(&self, chain: &impl BlockchainInterface) -> Result<u32, HeaderExtError>;
+
+    /// Returns the block's difficulty as a floating-point number.
+    fn get_difficulty(&self) -> f64;
+
+    /// Retrieves the height of the block in the blockchain.
+    ///
+    /// Returns an error if the block is not found in the chain.
+    fn get_height(&self, chain: &impl BlockchainInterface) -> Result<u32, HeaderExtError>;
+
+    /// Returns the block's target as a hexadecimal string.
+    fn get_target(&self) -> String;
+
+    /// Returns the block's version as a hexadecimal string.
+    fn get_version_hex(&self) -> String;
+}
+
+/// Represents errors that can occur when using the `HeaderExt` trait.
+#[derive(Debug)]
+pub enum HeaderExtError {
+    /// A generic error related to the blockchain interface.
+    Chain,
```

Comment:
I think this should wrap the actual error, we need to propagate up the actual chain error.